### ISSUE TITLE
Dev Task: add chat functionality

### ISSUE THREAD

**Issue by jonathaneeckhout** [jonathaneeckhout/jdungeon#24]
The player should be able to chat to other players.

There should be 3 levels of chat but for this milestone we can focus on the first one:

- Global chat -> broadcast a message to all players in a certain range
- Private chat -> a 1 to 1 communication channel
- Team chat -> only chat to your team members -> not needed now

There should also be a limit on how many messages can be send per minute to prevent spamming.

**Comment by NancokPS2** [comment]
Since it is an MMO, maybe these levels could be necessary:  
    

Global: to everyone playing, pretty much server only and reserved for sensitive announcements (it would be a good idea to keep it inside the chat system as just another level to reduce edge-cases)    
Broadcast: within the same zone (assuming we are implementing different areas to warp to). Can be used by players in a limited manner (trough some NPC or some other limited system)
Local: everyone within a certain range from the user (mostly other players on-screen)  
Private: 1 to 1  
Group: just like global, but tuned to only be received by a select amount of players
      
----------------------------------------  

We may be able to facilitate all of this with a filter system:  
Whenever a message is sent it also carries a series of tags, the server receives it and iterates trough every possible receiver, passing it onto them if their filter matches.  
  

For this we may want a MessageClient and MessageServer classes. The client simply sends a message to the server, while the server iterates trough every existent MessageClient and attempts to pass the message to valid ones.

**Comment by jonathaneeckhout** [comment]
Great summary.

If I may suggest a priority on what should be implemented first I would say:
Global, Local should be done first
Private next
Broadcast and group last (we don't have  group system yet).

For the local chat the actually network sync area can be reused. Every player inside the network sync area of the player typing the chat will receive the message. It's just a bit bigger than the screen.
